### ISSUE TITLE
[Bug]: preifx cache reuse

### ISSUE THREAD

**Issue by brotherchen** [vllm-project/vllm#9643]
### 🚀 The feature, motivation and pitch

I want to know whether the prefix cache in vllm supports automatic cache reuse before a request has been computed. For instance, if there is a new batch of requests with shared prefixes, will it automatically reuse the prefixes, or is it like TensorRT-LLM, where cache reuse for KV can only occur after the requestis complete?

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by brotherchen** [comment]
I closely studied the implementation and found that uncomputed blocks are assigned to the same batch of requests, but both requests need to fill their corresponding (the same) KV cache. Does this result in some degree of wasted computation, or is it a bug?

**Comment by rickyyx** [comment]
Yeah, I don't think right now prefix caching is optimal within the same batch for forward. (i.e. if multiple sequences share a common prefix, and the prefix has not been computed before, it will be computed multiple times while forwarding that batch). It's more of a limitation than bug I think. 

I think what you are referring to is the cascade inferencing feature - that's being worked on here. https://github.com/vllm-project/vllm/pull/8132/files

**Comment by brotherchen** [comment]
I do a simple but useful change on the block_manager_v2.
Presently, all request blocks are marked as computed only after the context length has been calculated. By updating the block state immediately after calculating the context length of each request, this issue can be resolved.

This optimization will yield benefits in certain scenarios without affecting the correctness of inference.
Orginal code:
![image](https://github.com/user-attachments/assets/ec6a44f5-b6f7-4f80-a8ff-7af9061b635e)


New code:
![image](https://github.com/user-attachments/assets/93639539-c83e-472d-ad02-144fc9486e5c)
